renderdoc/driver/vulkan/vk_counters.cpp

```diff
@@ -376,6 +514,197 @@ std::vector<CounterResult> VulkanReplay::FetchCountersAMD(const std::vector<GPUC
+      m_pReplay(rp),
+      m_QueryPool(qp)
+  {
+    m_pDriver->SetDrawcallCB(this);
+  }
+  ~VulkanKHRCallback() { m_pDriver->SetDrawcallCB(NULL); }
+
+  void PreDraw(uint32_t eid, VkCommandBuffer cmd) override
+  {
+    ObjDisp(cmd)->CmdBeginQuery(Unwrap(cmd), m_QueryPool, (uint32_t)m_Results.size(), 0);
+  }
+
+  bool PostDraw(uint32_t eid, VkCommandBuffer cmd) override
+  {
+    ObjDisp(cmd)->CmdEndQuery(Unwrap(cmd), m_QueryPool, (uint32_t)m_Results.size());
+    m_Results.push_back(eid);
+    return false;
+  }
+
+  void PostRedraw(uint32_t eid, VkCommandBuffer cmd) override {}
+  // we don't need to distinguish, call the Draw functions
+  void PreDispatch(uint32_t eid, VkCommandBuffer cmd) override { PreDraw(eid, cmd); }
+  bool PostDispatch(uint32_t eid, VkCommandBuffer cmd) override { return PostDraw(eid, cmd); }
+  void PostRedispatch(uint32_t eid, VkCommandBuffer cmd) override { PostRedraw(eid, cmd); }
+  void PreMisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd) override { PreDraw(eid, cmd); }
+  bool PostMisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd) override
+  {
+    return PostDraw(eid, cmd);
+  }
+  void PostRemisc(uint32_t eid, DrawFlags flags, VkCommandBuffer cmd) override
+  {
+    PostRedraw(eid, cmd);
+  }
+  void AliasEvent(uint32_t primary, uint32_t alias) override
+  {
+    m_AliasEvents.push_back(std::make_pair(primary, alias));
+  }
+
+  void PreEndCommandBuffer(VkCommandBuffer cmd) override {}
+  WrappedVulkan *m_pDriver;
+  VulkanReplay *m_pReplay;
+  VkQueryPool m_QueryPool;
+  std::vector<uint32_t> m_Results;
+  // events which are the 'same' from being the same command buffer resubmitted
+  // multiple times in the frame. We will only get the full callback when we're
+  // recording the command buffer, and will be given the first EID. After that
+  // we'll just be told which other EIDs alias this event.
+  std::vector<std::pair<uint32_t, uint32_t> > m_AliasEvents;
+};
+
+std::vector<CounterResult> VulkanReplay::FetchCountersKHR(const std::vector<GPUCounter> &counters)
+{
+  std::vector<uint32_t> counterIndices;
+  for (const auto& c : counters)
+    counterIndices.push_back(FromKHRCounter(c));
+
+  VkQueryPoolPerformanceCreateInfoKHR perfCreateInfo = {
+    VK_STRUCTURE_TYPE_QUERY_POOL_PERFORMANCE_CREATE_INFO_KHR,
+    NULL,
+    0,
+    (uint32_t)counterIndices.size(),
+    &counterIndices[0]
+  };
+  uint32_t passCount = 0;
+  ObjDisp(m_pDriver->GetInstance())->
+    GetPhysicalDeviceQueueFamilyPerformanceQueryPassesKHR(Unwrap(m_pDriver->GetPhysDev()),
+                                                          &perfCreateInfo,
+                                                          &passCount);
+
+  VkDevice dev = m_pDriver->GetDev();
+  VkAcquireProfilingLockInfoKHR acquireLockInfo = {
+    VK_STRUCTURE_TYPE_ACQUIRE_PROFILING_LOCK_INFO_KHR,
+    NULL, 0, 0 };
+  VkResult vkr =
+    ObjDisp(dev)->AcquireProfilingLockKHR(Unwrap(dev), &acquireLockInfo);
+  RDCASSERTEQUAL(vkr, VK_SUCCESS);
```

Comment:
I think we need to handle `vkAcquireProfilingLockKHR` being contended or unavailable here, it's not so unusual an error condition that we can ignore it and crash with no reasonable recovery.

I think we should set some (small) timeout like 50-100ms, and if it still fails then try to bail out of this fetch of counter data with empty data.

Reply:
Done.